### PR TITLE
auto-update: support pods

### PR DESCRIPTION
**Pull request by vrothberg**
Support auto updating containers running inside pods.  Similar to
containers, the systemd units need to be generated via
`podman-generate-systemd --new $POD` to generate the pod's units.
 
Note that auto updating a container inside a pod will restart the entire
pod.  Updates of multiple containers inside a pod are batched, such that
a pod is restarted at most once.  That is effectively the same mechanism
for auto updating containers in a K8s YAML via the `podman-kube@`
template or via Quadlet.
 
Updating a single container unit without restarting the entire pod is
not possible.  The reasoning behind is that pods are created with
--exit-policy=stop which will render the pod to be stopped when auto
updating the only container inside the pod.  The (reverse) dependencies
between the pod and its containers unit have been carefully selected for
robustness.  Changes may entail undesired side effects or backward
incompatibilities that I am not comfortable with.
 
Fixes: #17181
Signed-off-by: Valentin Rothberg <vrothberg@redhat.com> 

#### Does this PR introduce a user-facing change?

```release-note
Support auto updates for containers running inside a pod.
```

### CODE REVIEW

**Review by Luap99 — COMMENTED** [review]
Can you add a test for it?

Interestingly the units I have deployed on my home server do not contain the BindsTo=, I do not remember why I changed it but I think I likely wanted to restart a single container without the pods so this change looks good to me.
Although I wonder if this could break some user expectations, I think at least you release note message should be more clear because and mention the exact change.

**Comment by vrothberg** [comment]
> Can you add a test for it?

I hoped I could sneak it in but I will have a look :)  The restart should be easier to test.

> Interestingly the units I have deployed on my home server do not contain the BindsTo=, I do not remember why I changed it but I think I likely wanted to restart a single container without the pods so this change looks good to me. Although I wonder if this could break some user expectations, I think at least you release note message should be more clear because and mention the exact change.

I was thinking about it but fortunately found the get-free-out-of-jail-ticket in the docs:
> Note  systemctl should only be used on the pod unit and one should not start or stop containers individually via systemctl, as they are managed by the pod service along with the internal infra-container.

So we never gave a guarantee on the reverse dependency being stable.  Unfortunately, I don't see another way to fix the auto update issue.

**Comment by Luap99** [comment]
Yes I am totally on board with this change but the release note should be more clear,
how about: 
```
Removed the `BindsTo=` option from the generated systemd units to support auto updating individual containers in pods with podman auto-update
```

**Comment by rhatdan** [comment]
LGTM

**Comment by vrothberg** [comment]
I had to revise the initial version of the PR where I only relaxed the reverse dependencies.  Explanation is in the PR/commit message.

@Luap99 PTAL - as discussed yesterday.

**Comment by vrothberg** [comment]
@edsantiago PTAL

File: test/system/255-auto-update.bats [hunk 1]
```diff
@@ -525,4 +525,72 @@ EOF
...
+    run_podman pod inspect --format "{{.ID}}" $podname
+    assert "$output" != "$podid" "pod has been recreated"
+    run_podman container inspect --format "{{.ID}}" $ctrname
+    assert "$output" != "$ctrid" "container has been recreated"
+
+    run systemctl stop pod-$podname.service
+    assert $status -eq 0 "Error stopping pod systemd unit: $output"
+
+    run_podman pod rm -f $podname
+    run_podman rmi $local_image
```
Comment (edsantiago) [inline test/system/255-auto-update.bats:591]: LGTM but please also `rmi $(pause_image)` to avoid nasty red test-log clutter

Comment (edsantiago) [reply test/system/255-auto-update.bats:591]: (There's other nasty red clutter I need to go fix, so if you want to merge without fixing this, it's fine with me, I'll come in later today and clean up)

**Comment by vrothberg** [comment]
@containers/podman-maintainers PTAL

File: test/system/255-auto-update.bats (hunk 1 above)
Comment (vrothberg) [reply test/system/255-auto-update.bats]: I don't see the other one :( Thanks for cleaning up after me!

**Review by Luap99 — COMMENTED** [review]
I rather not change the PODMAN_SYSTEM_UNIT env var to point to a different unit. This will break my passthrough logs progress. Also hard coding the name will break portability.

Instead I would teach podman auto-update to look if in pod then read the env from the pod and use this name to restart.

**Comment by vrothberg** [comment]
>  Also hard coding the name will break portability.

The dependencies are already hard-coded, so I don't think that's an issue.

> This will break my passthrough logs 

That's a very fair point.  Updating the auto-update code to use the pod's unit instead sounds like a nice alternative :+1:

**Comment by vrothberg** [comment]
@Luap99 PTanotherL.  It was fairly straight-forward.

**Review by Luap99 — COMMENTED** [review]
LGTM

**Comment by edsantiago** [comment]
I'm seeing a test failure now, `image is in use by container`

**Comment by vrothberg** [comment]
> I'm seeing a test failure now, `image is in use by container`

Can you share which test is failing on your end?

**Comment by vrothberg** [comment]
Infra is sick atm.

**Comment by edsantiago** [comment]
I can't replicate; must've been a screwup on my end. And yes, CI is unwell.

**Comment by edsantiago** [comment]
rebase on main to get CI to pass. (Sorry, fix is only on main, not side branches)

**Comment by vrothberg** [comment]
Thanks, rebased :heavy_check_mark:

**Comment by vrothberg** [comment]
Ready to merge

**Pull request merged**